hasktorch/src/Torch/Typed/Functional.hs

```diff
@@ -1191,8 +1191,31 @@ transpose2D
   -> Tensor device dtype '[j, i] -- ^ output
 transpose2D = transpose @0 @1
 
--- diag :: Tensor device dtype shape -> Int -> Tensor device dtype shape
--- diag t index = unsafePerformIO $ (ATen.cast2 ATen.Managed.tensor_diag_l) t index
+type family DiagShape (index :: Nat) (shape :: [Nat]) :: [Nat] where
+  DiagShape i '[n] = '[n + i, n + i] -- 1d -> 2d case
+  DiagShape i '[m, n] = '[Min m n - i] -- 2d -> 1d case
+  DiagShape _ _ =  TypeError (Text "The input must be a matrix or a vector.")
```

Comment:
that's a great addition and will help the user, thanks!
```suggestion
  DiagShape _ shape =  TypeError (Text "The input must be a matrix or a vector, but it has " :<>: ListLength shape  :<>: " dimensions.")
```
if you can, you may also want to guard against invalid `i`s :)